### ISSUE TITLE
Need a better solution for updating DOM when user changes source

### ISSUE THREAD

**Issue by jhpedemonte** [maqetta/maqetta#1522]
Currently, when a user updates the source of a file, Maqetta redoes the whole DOM in the VE iframe.  This can lead to issues, though, since the JS context isn't also destroyed (see main discussion in #1511; also #1520).

---

First, if redoing the entire VE DOM, we should also destroy the existing JS context.  Most likely, this means destroying and recreating the iframe.

However/ recreating the iframe may have performance implications.  Therefore, we should look for solutions that allow us to update only a part of the DOM if only a small part of the source has changed.

@peller @JonFerraiolo @waynevicknair

**Comment by peller** [comment]
I think we should first prove that recreating the iframe is not feasible

**Comment by jhpedemonte** [comment]
I don't think we want to recreate the iframe on every source change.  That's bad technique.

Rather, we should do some quick calculations when the source gets updated:
- Was the source **change contained within a single widget**?  Then only update that widget in the model (& VE).
- Did the source **change affect a subset of widgets**?  Then only update those widgets.
- Did the source **change affect the `<head>`**?  Then here we probably have to recreate the iframe, since the JS or CSS contexts may have changed.
  - Although we could also be fine grained about this as well.  For example, was the change the addition of a resource?  Then only load that resource in model/VE.

An quick (but not complete) solution for the first 2 bullet points would be to see if the **change was limited to the `<body>`**, in which case we do not have to worry about refreshing/recreating the iframe.

**Comment by JonFerraiolo** [comment]
@jhpedemonte: Hallelujah to your comments. I expect that optimizing the update logic when changing things in the source editor will be one of the most important performance enhancements we will make in the coming months.

Note that if we try to update only a subset of the document (e.g., a particular widget), we might run into various rare and surprising bugs because in web pages things are sometimes surprisingly interconnected. For example, changing the source for a DataGrid (or other data-connected widget) might require updates to more of the visual editor than just the DataGrid. Just brainstorming, but perhaps we can have a widget-specific metadata flag that we can put on certain widgets to say that source updates require more global recalculations in the visual editor.

**Comment by peller** [comment]
Either way, we have tons of optimization to do when the iframe is reloaded, like the tokenizer.  Seems like we ought to start there.

**Comment by peller** [comment]
...or consider whether we should be using our modeler at all, or leverage whatever Orion is doing.

**Comment by jhpedemonte** [comment]
-> M6

**Comment by peller** [comment]
See also #1554

**Comment by JonFerraiolo** [comment]
-> @jhpedemonte M7 Prio-Med (I assume @jhpedemonte will reassign)

**Comment by jhpedemonte** [comment]
-> M9
